specification/2.0/README.md

```diff
@@ -469,7 +469,7 @@ The following example defines two buffer views: the first is an ELEMENT_ARRAY_BU
 }
 ```
 
-When a buffer view is used for vertex attribute data, it may have a `byteStride` property. This property defines the stride in bytes between each vertex.
+When a buffer view is used for vertex attribute data, it may have a `byteStride` property. This property defines the stride in bytes between each vertex. Buffer views with other types of data should not define `byteStride`.
```

Comment:
See https://github.com/KhronosGroup/glTF/issues/1885 for context here, but could EXT_mesh_gpu_instancing also use interleaved attributes? I think the answer is yes, and that `bufferView.target` would similarly be `ARRAY_BUFFER`, but I'm not sure about this.

Reply:
I'm not sure how to generalize and future-proof this language. Perhaps we should be describing "attribute data" to cover both vertex attributes and instance attributes?

Reply:
> bufferView.target would similarly be ARRAY_BUFFER

I don't think ARRAY_BUFFER is right... It would be if the implementation is using vertex stream divisors to implement instancing, but using uniform or shader storage or texel buffers would be a more modern way of achieving this, for which ARRAY_BUFFER is incorrect. So I'd rather have a different type of target that we expect for the instance data stream...

Reply:
Effectively I'd say the rule should be that the target defines the *usage* of the data, but not the implementation details of a given renderer. Ideally we wouldn't use GL names for this but I guess that ship has sailed.

Reply:
Fine with me to not use `ARRAY_BUFFER` for instance attributes. However, I expect we do want to allow instance attributes to be interleaved, right? Does gltfpack interleave these attributes? If so we would want to allow `.byteStride` in that use case.

Reply:
gltfpack doesn't interleave attributes at the moment.

For whether interleaving is allowed in general on a given data type, I think we should look at the implementation complexity (where support for byte stride is easy when the data is mapped directly to a vertex stream on the GPU, and more difficult otherwise), possible gains in efficiency from interleaving due to increased cache coherency, and uniformity.

Instanced data is interesting because it has the widest possible set of implementation option out of all other types of data:

- It can be implemented by uploading data to the GPU and using vertex streams with instance frequency/divisor; in this case supporting interleaving is free.
- It can be implemented by uploading data to the GPU and using buffers (typed or untyped) with the shader reading data from the buffer based on gl_InstanceId or equivalent; in this case, interleaving can be difficult and I'd expect implementations to have to pre-process interleaved data on the CPU
- It can be implemented by reading the transforms on the CPU and dispatching multiple draw calls (which can still be more efficient vs non-instanced implementation for a few reasons); in this case, interleaving might require extra processing on the CPU side

I don't think there are efficiency gains that one should expect from interleaving the instance data however, since the model is generally one where you'd need to read all transforms to render the batch of instances. One could make the argument that for animations, interleaving is more valuable than it is for instances since cache utilization can be better when keyframes are ordered globally by time across multiple different tracks, so that sampling animation at time T requires fewer than 1 cache accesses per track.

If the spec *right now* only allows byteStride to be defined for vertex data (which I think was the intention), maybe it would make sense to phrase this as "Buffer views with other types of data should not define `byteStride`, unless interleaved data is explicitly supported by the extension that uses the buffer view as a data source"? That way we can have specific callouts to byteStride being supported.

Alternatively I can remove this specific note from the change, if it's not clear whether we should support interleaving for non-vertex data.

Reply:
> If the spec right now only allows byteStride to be defined for vertex data (which I think was the intention)...

I guess this is what we need to determine, then... let me follow up in https://github.com/KhronosGroup/glTF/issues/1864.

Reply:
fbx2gltf currently uses target=0 for all morph target data buffer views. I noticed gltfpack uses 34962 ( vertex data ). I ran into this exact contradiction around byteStride when packing morph target data without specifying a target for the buffer view. Perhaps this is an error on fbx2gltf's part and all morph target data should be changed to use 34962 target type.

We use GLTF/GLB to deliver assets to clients, who then process them based on the client they use (Unity, Unreal, etc). None of our clients can upload the GLB data directly into the GPU and have to process it for the particular engine. We would like to see GLTF/GLB be as flexible as possible when transmitting data to clients.

Reply:
> fbx2gltf currently uses target=0 for all morph target data buffer views
Does it specify byteStride for these or not? If it doesn't then this change shouldn't impact that in any way, as there's no requirement in the existing spec that the vertex views must specify non-0 target.

Reply:
I would like to include indices in the tightened specification, one reason for this is consistency - both type of bufferViews will be read by the GPU in the vertex/geometry processing step(s).
Another reason is with regards to component type:

`When a buffer view contain vertex indices or attributes, they must be its only content, i.e., it's invalid to have more than one kind of data in the same buffer view.`

Ie, it should not be valid to have 2 Accessors (indices) referencing the same BufferView, using different offsets and 0 byteStride to access the data using different component types (short/int)
At the moment it seems the validator does not flag for this.

Reply:
> Effectively I'd say the rule should be that the target defines the semantics of the data, but not the implementation details of a given renderer.

I totally agree and for me it would help if target _must_ be specified if the BufferView is used for data that needs to be bound (in a buffer before being accessed by the GPU).
That way an implementations can look through the declared BufferViews and know which ones will require a GPU based target buffer.
Ideally this could be expanded to other targets so that implementations may decide what data to generate buffers for (without traversing the nodetree and check what will be rendered)

Reply:
Since today index buffer views can't contain byteStride, I don't think it's a good idea to include that as that'd render many existing assets invalid. Ditto for the target - I agree it should have been a required attribute, but since it's not it's not an option to go back.

Reply:
> Ie, it should not be valid to have 2 Accessors (indices) referencing the same BufferView, using different offsets and 0 byteStride to access the data using different component types (short/int)

I don't think I understand this part — we wouldn't want to interleave indices, but what's the concern with multiple indices' accessors being end-to-end in a single buffer view?

Reply:
There is also the rule about byteStride being required when two or more accessors use the same buffer view. 
(https://github.com/KhronosGroup/glTF/issues/1198). 

I ran into this issue because our morph targets use target=0 in the buffer views, so we ran into this valiadtion error when we pack the morph data into the same buffer view, even thought it is tightly packed and the stride equals FLOAT, VEC3, since the morph pos/nor/tan are are the same componentType and Type.

But, there is also the rule that the skinning matricies cannot have a byte stride, which also uses target=0 in the buffer view.

I had to add extra code to detect when I'm packing accessors into the same buffer view to specify a byte stride or not based on if it is vertex data, index data, or 'other' data, but not 'other' data that is bind matricies.

Reply:
I'm not sure what `target=0` means sorry, that's not a valid enum value here. Do you mean `target` is undefined?

Only vertex attributes may have a byte stride, and we can't loosen that safely within the 2.0 version (except perhaps in extensions that define new types of data that might benefit from a byteStride), so I think the purpose of this subthread is resolved. If there's something else I've missed in this thread, let's open a new PR or issue to continue its discussion.